apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java

```diff
@@ -0,0 +1,144 @@
+import com.apollocurrency.aplwallet.apl.core.db.derived.DerivedTableInterface;
+import com.apollocurrency.aplwallet.apl.core.shard.ShardConstants;
+import com.apollocurrency.aplwallet.apl.core.shard.ShardPresentData;
+import com.apollocurrency.aplwallet.apl.core.shard.helper.CsvImporter;
+import com.apollocurrency.aplwallet.apl.util.Zip;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Interface for shard data downloading management. It does following:
+ * In case empty database (node is started from the scratch) it will do:
+ * - connect to peers and check if sharding is present in current network
+ * - gather statistics info about shard hashes/data from peers
+ *
+ * If shard is not present it starts 'old' process importing Genesis data
+ */
+@Slf4j
+@Singleton
+public class ShardDownloadPresenceObserver {
+
+    private DatabaseManager databaseManager;
+    private Blockchain blockchain;
+    private BlockchainProcessor blockchainProcessor;
+    private DerivedTablesRegistry derivedTablesRegistry;
+    private CsvImporter csvImporter;
+    private Zip zipComponent;
+
+    @Inject
+    public ShardDownloadPresenceObserver(DatabaseManager databaseManager, BlockchainProcessor blockchainProcessor,
+                                         Blockchain blockchain, DerivedTablesRegistry derivedTablesRegistry,
+                                         Zip zipComponent, CsvImporter csvImporter) {
+        this.databaseManager = Objects.requireNonNull(databaseManager, "databaseManager is NULL");
+        this.blockchainProcessor = Objects.requireNonNull(blockchainProcessor, "blockchainProcessor is NULL");
+        this.derivedTablesRegistry = Objects.requireNonNull(derivedTablesRegistry, "derivedTablesRegistry is NULL");
+        this.blockchain = Objects.requireNonNull(blockchain, "blockchain is NULL");
+        this.zipComponent = Objects.requireNonNull(zipComponent, "zipComponent is NULL");
+        this.csvImporter = Objects.requireNonNull(csvImporter, "csvImporter is NULL");
+    }
+
+    /**
+     * Method is called when node knows it's int sharded network and shard ZIP information has been downloaded by transport subsystem.
+     *
+     * @param shardPresentData shard present data contains downloaded ZIP name
+     */
+    public void onShardPresent(@ObservesAsync @ShardPresentEvent(ShardPresentEventType.PRESENT) ShardPresentData shardPresentData) {
+        // shard archive data has been downloaded at that point and stored (unpacked?) in configured folder
+        String zipFileName = shardPresentData.getFileIdValue();
+        Path zipInFolder = csvImporter.getDataExportPath().resolve(zipFileName + ".zip").toAbsolutePath();
+        log.debug("Try unpack file name '{}'", zipInFolder);
+        boolean unpackResult = zipComponent.extract(zipFileName + ".zip", csvImporter.getDataExportPath().toString());
+        log.debug("Zip is unpacked = {}", unpackResult);
+        Genesis.apply(); // import public data - genesis public Keys + balances
```

Comment:
We should import only genesis_public_key

Reply:
Done